### PR TITLE
Initial commit of transcoding.

### PR DESCRIPTION
**Pull request by tokejepsen**
This PR handles splitting the encoding workload of a sequence of shots to multiple machines.

This transcoding framework requires Deadline, OIIO (if transcoding EXRs with DWAA compression) and two families for the standalone publisher is needed (https://github.com/pypeclub/pype-config/blob/develop/presets/standalone_publish/families.json);

```json
    "create_transcode_prores_422_hq": {
      "name": "prores422HQ",
      "label": "Transcode - Prores 422 HQ",
      "family": "prores422HQ",
      "icon": "image",
      "defaults": ["Main"],
      "help": "Editorial files to transcode to Prores 422 HQ."
    },
    "create_transcode_h264": {
      "name": "h264",
      "label": "Transcode - H264",
      "family": "h264",
      "icon": "image",
      "defaults": ["Main"],
      "help": "Editorial files to transcode to H264."
    },
    "create_transcode_prores_proxy": {
      "name": "proresProxy",
      "label": "Transcode - Prores Proxy",
      "family": "proresProxy",
      "icon": "image",
      "defaults": ["Main"],
      "help": "Editorial files to transcode to Prores Proxy (half resolution)."
    }
```

Workflow:

1. Generate an editorial file (xml/otio) and an audio file (wav). These two files need to be located in the same folder on disk.
2. Using the standalone publisher you publish the editorial file, using one of the currently supported destination presets;  prores422HQ, proresProxy or H264.
3. When publishing, a job is sent to Deadline for each shot found in the editorial file and one to concatenate all the shots together to one movie file. All shots and sequence movie files are rendered next to the editorial file.

Its currently only tested with xml/otio editorial file with a single track of whole clips (no trim). Because concatenation of video files with audio can cause frozen frames (https://trac.ffmpeg.org/ticket/9052), all shot transcodes are without audio. The audio from the publish is added to the concatenation command so the sequence gets the audio.

Further improvements could be;
- Ftrack review upload/publish.
- Support more output presets.
- Support more input editorial file types.
- Audio on shot transcodes.
- Optional clean up of temporary shot transcodes.
- Max length of shot, to split the encoding work load of long shots.
- DWAA/DWAB compressed exr compatibility with ffmpeg code refactor (should already be in Pype).
- Support more render farm software.
- Optimize render times: 
    - If a shots movie file is the same codec as the destination sequence, then copying the video stream would be quicker.
    - If outputting multiple codecs (this can only be determined at the publishing stage) of that same edit, it'll be faster to only decode DWAA EXRs once.
- Support more editorial effects such as overlays, dissolves and retimes.
- Support timecode and burnins.
- Validate length of audio and edit sequence to get the same (currently shortest is the determining).
- Transcoding from Editorial applications. In NukeStudio we could tag items, video track and sequences for transcoding. The tag would be destination codec/format.

### CODE REVIEW

**Comment by tokejepsen** [comment]
As a side note, we are utilizing Ftrack to find the latest versions of certain stages of the production, which we can generate an editorial file from. This makes is possible for anyone with a little knowledge to generate movie files of the latest state of a production.

**Comment by mkolar** [comment]
Tons of potential with this. Thanks a lot we'll have a look properly

**Comment by tokejepsen** [comment]
When available this PR could do with an updated OTIO for the image sequences. Currently a bit fiddly to get the correct ranges.

**Comment by tokejepsen** [comment]
> When available this PR could do with an updated OTIO for the image sequences. Currently a bit fiddly to get the correct ranges.

Getting for correct start frame for image sequences is not easy, so updated OTIO is definitely needed for wide spread production use.

File: pype/scripts/transcode.py
Comment (jezscha) [inline pype/scripts/transcode.py:59]: The output_path is used as an argument, but it is immediately after overridden by other path. And truly it is written to the staging directory.

File: pype/scripts/transcode.py
Comment (jezscha) [inline pype/scripts/transcode.py:74]: did you mean this? 
```
output_path = output_path or os.path.join(
    output_path, name + preset_templates[preset]["extension"]
).replace("//", "/")
```

**Review by jezscha — CHANGES_REQUESTED** [review]

**Comment by jezscha** [comment]
Also in description you are using this family preset: 

```json
{
	"create_transcode_prores_proxy": {
      "name": "proresProxy",
      "label": "Transcode - Prores Proxy",
      "family": "proresProxy",
      "icon": "image",
      "defaults": ["Main"],
      "help": "Editorial files to transcode to Prores Proxy (half resolution)."
    }
}
```

But there is nothing in code to support this.

File: pype/plugins/standalonepublisher/publish/collect_transcode.py
Comment (tokejepsen) [reply pype/plugins/standalonepublisher/publish/collect_transcode.py:17]: Good point. Think this is a left over from previous iterations. Will remove.

File: pype/plugins/standalonepublisher/publish/submit_transcode_deadline.py
Comment (tokejepsen) [reply pype/plugins/standalonepublisher/publish/submit_transcode_deadline.py:50]: Think it was to be explicit in the modules, to prevent implicit modules messing up the job on the farm. Good point about the `clique` module. Will add it.

File: pype/plugins/standalonepublisher/publish/submit_transcode_deadline.py
Comment (tokejepsen) [reply pype/plugins/standalonepublisher/publish/submit_transcode_deadline.py:51]: Sorry, did you want me to commit the above code?

File: pype/scripts/transcode.py
Comment (tokejepsen) [reply pype/scripts/transcode.py:59]: The output path is root directory for all the shots, so before being used in the ffmpeg command, it'll need to be modified with the file name and extension.
Do you want it different somehow? Maybe call it root_path or something?

File: pype/scripts/transcode.py
Comment (tokejepsen) [reply pype/scripts/transcode.py:74]: See above comment on `output_path` argument.

**Comment by tokejepsen** [comment]
> Also in description you are using this family preset:
> 
> ```json
> {
> 	"create_transcode_prores_proxy": {
>       "name": "proresProxy",
>       "label": "Transcode - Prores Proxy",
>       "family": "proresProxy",
>       "icon": "image",
>       "defaults": ["Main"],
>       "help": "Editorial files to transcode to Prores Proxy (half resolution)."
>     }
> }
> ```
> 
> But there is nothing in code to support this.

Sorry, you are right. Missing a commit for this. Will add now.

File: pype/plugins/standalonepublisher/publish/submit_transcode_deadline.py
Comment (jezscha) [reply pype/plugins/standalonepublisher/publish/submit_transcode_deadline.py:52]: up to you @tokejepsen we can change it on our side before merge if you want.

File: pype/scripts/transcode.py
Comment (jezscha) [reply pype/scripts/transcode.py:59]: I see, I was thinking it should be used as target for files to be exported to. So if your context in standalonepublisher is at folder `shots` in the project, then files will be exported there. But it was misunderstood.

File: pype/plugins/standalonepublisher/publish/submit_transcode_deadline.py
Comment (tokejepsen) [reply pype/plugins/standalonepublisher/publish/submit_transcode_deadline.py:52]: Ok, will leave it up to you guys.

File: pype/scripts/transcode.py
Comment (tokejepsen) [reply pype/scripts/transcode.py:59]: Currently its not referencing any published paths. All transcoded files get dumped to the same folder as the editorial file.

**Comment by mkolar** [comment]
I think we can merge this for good measure, as we're converting it to 3.x anyways. No harm done in having it in the old branches for historical reasons

**Review by mkolar — APPROVED** [review]

**Comment by tokejepsen** [comment]
Might be good to have the "Further Improvements" as an issue to tackle in the future?

**Comment by mkolar** [comment]
@tokejepsen agreed. those should be converted to a bigger issue, so we don't forget. We'll soon be starting a major chunk of work on transcoding and reviewables in general, so having it for posterity would be great.

**Pull request merged**
